# iNaturalist status updates by state - QLD

Using the file produced in the collate-status-taxa.ipynb: `inat-aust-status-taxa.csv`, generate lists to update iNaturalist statuses

**Next steps:**
Establish the changes that need to be made. Read in the sensitive and conservation list for each state.
    a. new - any new species that appear in the state lists but do not have a status in inaturalist (new template)
    b. updates - any changes to information which was added by us previously (user_id = 708886) (update template, action='UPDATE')
    c. removals - any statuses which were added by us previously (user_id = 708886) list which are incorrect (update template, action='REMOVE')
    d. flags - are there any statuses by other users that need to be flagged?

## Prep - common to all states
1. Read in the inaturalist statuses & filter out this state
2. Read in the inaturalist taxa list
3. Read in the state sensitive and conservation list
4. Attempt to match the state statuses to an IUCN equivalent
5. Determine the best placeID to use for this state


### 1. iNaturalist statuses

In [1]:
import pandas as pd

# projectdir = "/Users/oco115/PycharmProjects/authoritative-lists/" # basedir for this gh project
projectdir = "/Users/new330/IdeaProjects/authoritative-lists/" # basedir for this gh project
sourcedir = projectdir + "source-data/inaturalist-statuses/"
listdir = projectdir + "current-lists/"


# read in the statuses
taxastatus = pd.read_csv(sourcedir + "inat-aust-status-taxa.csv", encoding='UTF-8',na_filter=False,dtype=str) ## Read inaturalist conservation statuses file
taxastatus.head(3)

id taxon_id  user_id place_id source_id  \
0  166449    38493  1138587     7830             
1  234788   918383   702203     9994             
2  234789   918383   702203     7308             

                            authority status  \
0  Flora and Fauna Guarantee Act 1988     CR   
1           Atlas of Living Australia    NT    
2           Atlas of Living Australia     LC   

                                                 url description geoprivacy  \
0                                                                  obscured   
1  https://bie.ala.org.au/species/https://id.biod...                          
2  https://bie.ala.org.au/species/https://id.biod...                          

   ...         genus specificEpithet infraspecificEpithet  \
0  ...     Eulamprus      kosciuskoi                        
1  ...  Chiloschista     phyllorhiza                        
2  ...  Chiloschista     phyllorhiza                        

               modified            scientificName taxonRank  \
0  2021-03-01T10:35:01Z      Eulamprus kosciuskoi   species   
1  2022-01-08T03:30:36Z  Chiloschista phyllorhiza   species   
2  2022-01-08T03:30:36Z  Chiloschista phyllorhiza   species   

                                          references preferred_common_name  \
0  http://reptile-database.reptarium.cz/search.ph...                         
1  http://www.catalogueoflife.org/annual-checklis...                         
2  http://www.catalogueoflife.org/annual-checklis...                         

  is_active current_synonymous_taxon_ids  
0                                         
1                                         
2                                         

[3 rows x 34 columns]

In [2]:
def filter_state_statuses(stateregex: str, urlregex: str):
    authoritydf = taxastatus['authority'].drop_duplicates().sort_values()
    authoritydf = authoritydf[pd.Series(authoritydf).str.contains(stateregex)]
    urldf = taxastatus['url'].drop_duplicates().sort_values()
    urldf = urldf[pd.Series(urldf).str.contains(urlregex)]
    placedisplaydf = taxastatus['place_display_name'].drop_duplicates().sort_values()
    placedisplaydf = placedisplaydf[pd.Series(placedisplaydf).str.contains(stateregex)]
    placedf = taxastatus['place_name'].drop_duplicates().sort_values()
    placedf = placedf[pd.Series(placedf).str.contains(stateregex)]
    # concat all and remove duplicates
    statedf = pd.concat([taxastatus.apply(lambda row: row[taxastatus['place_display_name'].isin(placedisplaydf)]),
                         taxastatus.apply(lambda row: row[taxastatus['place_name'].isin(placedf)]),
                         taxastatus.apply(lambda row: row[taxastatus['url'].isin(urldf)]),
                         taxastatus.apply(
                             lambda row: row[taxastatus['authority'].isin(authoritydf)])]).drop_duplicates()
    return statedf.sort_values(['taxon_id', 'user_id'])


inatstatuses = filter_state_statuses(" WA |WEST AUST|West Aust|WESTERN AUSTRALIA|Western Australia", ".wa.gov.au")
inatstatuses.rename(columns={'id':'status_id','id_y':'taxon_id_y'},inplace=True)
inatstatuses

status_id taxon_id user_id place_id source_id  \
1990    153386   100948  708886     6827     16654   
2158    153585   101164  708886     6827     16654   
1982    153373   101165  708886     6827     16654   
2346    153802   101474  708886     6827     16654   
2019    153420   101509  708886     6827     16654   
...        ...      ...     ...      ...       ...   
2311    153762    99344  708886     6827     16654   
2039    153443    99634  708886     6827     16654   
2185    153613    99973  708886     6827     16654   
2293    153742    99974  708886     6827     16654   
2171    153597    99977  708886     6827     16654   

                                           authority                 status  \
1990  WA Department of Environment and Convservation             endangered   
2158  WA Department of Environment and Convservation             vulnerable   
1982  WA Department of Environment and Convservation             endangered   
2346  WA Department of Environment and Convservation             endangered   
2019  WA Department of Environment and Convservation             endangered   
...                                              ...                    ...   
2311  WA Department of Environment and Convservation             endangered   
2039  WA Department of Environment and Convservation             endangered   
2185  WA Department of Environment and Convservation  critically endangered   
2293  WA Department of Environment and Convservation             endangered   
2171  WA Department of Environment and Convservation             endangered   

                                                    url description  \
1990  https://lists.ala.org.au/speciesListItem/list/...               
2158  https://lists.ala.org.au/speciesListItem/list/...               
1982  https://lists.ala.org.au/speciesListItem/list/...               
2346  https://lists.ala.org.au/speciesListItem/list/...               
2019  https://lists.ala.org.au/speciesListItem/list/...               
...                                                 ...         ...   
2311  https://lists.ala.org.au/speciesListItem/list/...               
2039  https://lists.ala.org.au/speciesListItem/list/...               
2185  https://lists.ala.org.au/speciesListItem/list/...               
2293  https://lists.ala.org.au/speciesListItem/list/...               
2171  https://lists.ala.org.au/speciesListItem/list/...               

     geoprivacy  ...          genus specificEpithet infraspecificEpithet  \
1990   obscured  ...  Fibulacamptus       bisetosus                        
2158   obscured  ...     Galaxiella           munda                        
1982   obscured  ...     Galaxiella    nigrostriata                        
2346   obscured  ...    Glacidorbis    occidentalis                        
2019   obscured  ...        Glyphis        garricki                        
...         ...  ...            ...             ...                  ...   
2311   obscured  ...  Diplodactylus        capensis                        
2039   obscured  ...    Dupucharopa    millestriata                        
2185   obscured  ...                                                       
2293   obscured  ...        Engaewa         reducta                        
2171   obscured  ...        Engaewa        walpolea                        

                  modified            scientificName taxonRank  \
1990  2021-10-28T19:35:25Z   Fibulacamptus bisetosus   species   
2158  2019-11-23T07:13:44Z          Galaxiella munda   species   
1982  2019-11-23T07:13:25Z   Galaxiella nigrostriata   species   
2346  2021-10-29T17:47:10Z  Glacidorbis occidentalis   species   
2019  2019-04-18T19:04:37Z          Glyphis garricki   species   
...                    ...                       ...       ...   
2311  2018-11-17T23:47:54Z    Diplodactylus capensis   species   
2039  2021-10-29T15:16:07Z  Dupucharopa millestriata   species   
2185                           Engaewa pseudore

### 2. iNaturalist taxonomy

In [3]:
# Output files contain these fields
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# so we need to match species from the state lists to the inat taxa to get the taxon_id

import zipfile
url = "https://www.inaturalist.org/taxa/inaturalist-taxonomy.dwca.zip"
filename = url.split("/")[-1]

z=zipfile.ZipFile(sourcedir + filename)

with z.open('taxa.csv') as from_archive:
    inattaxa = pd.read_csv(from_archive,dtype=str)
z.close()
inattaxa.head(3)


id                             taxonID                          identifier  \
0  1  https://www.inaturalist.org/taxa/1  https://www.inaturalist.org/taxa/1   
1  2  https://www.inaturalist.org/taxa/2  https://www.inaturalist.org/taxa/2   
2  3  https://www.inaturalist.org/taxa/3  https://www.inaturalist.org/taxa/3   

                         parentNameUsageID   kingdom    phylum class order  \
0   https://www.inaturalist.org/taxa/48460  Animalia       NaN   NaN   NaN   
1       https://www.inaturalist.org/taxa/1  Animalia  Chordata   NaN   NaN   
2  https://www.inaturalist.org/taxa/355675  Animalia  Chordata  Aves   NaN   

  family genus specificEpithet infraspecificEpithet              modified  \
0    NaN   NaN             NaN                  NaN  2021-11-02T06:05:44Z   
1    NaN   NaN             NaN                  NaN  2021-11-23T00:40:18Z   
2    NaN   NaN             NaN                  NaN  2022-12-27T07:33:16Z   

  scientificName taxonRank                                         references  
0       Animalia   kingdom  http://www.catalogueoflife.org/annual-checklis...  
1       Chordata    phylum  http://www.catalogueoflife.org/annual-checklis...  
2           Aves     class  http://www.catalogueoflife.org/annual-checklis...

### 3. State lists

Get the ALA Sensitive list: `geoprivacy` = `obscured`
Everything on the WA list is sensitive


In [ ]:
%%script echo skipping # comment this line to download dataset from lists.ala.org.au the web and save locally

import sys
import os
sys.path.append(os.path.abspath(projectdir + "source-code/includes"))
import list_functions as lf
alasensitivelist = lf.download_ala_list("https://lists-test.ala.org.au/ws/speciesListItems/dr18406?max=10000&includeKVP=true")
alasensitivelist = lf.kvp_to_columns(alasensitivelist)
alasensitivelist.to_csv(sourcedir + "wa-ala.csv")

Use the GBIF names parser to clean up the names

In [79]:
%%script echo skipping # comment this line to run the gbif parser again the web and save a file locally

import requests

namesonly = alasensitivelist['name']
url = "https://api.gbif.org/v1/parser/name"
headers = {'content-type' : 'application/json'}
data = namesonly.to_json(orient="values")
params = {'name':data}
r = requests.post(url=url,data=data,headers=headers)
results = pd.read_json(r.text)
results.to_csv(sourcedir + "wa-gbif.csv")
results

In [105]:
alasensitivelist = pd.read_csv(sourcedir + "wa-ala.csv", dtype=str)
parsednames = pd.read_csv(sourcedir + "wa-gbif.csv", dtype=str)
alasensitivelist = alasensitivelist.merge(parsednames[['scientificName','canonicalName']],how="left",left_on="name",right_on="scientificName")
alasensitivelist['wa_geoprivacy'] = 'obscured'
alasensitivelist['waTaxonID'] = alasensitivelist['W A Taxon Id']#.apply(lambda x: int(float(x)))
alasensitivelist['scientificName'] = alasensitivelist['canonicalName']
alasensitivelist['status'] = alasensitivelist['sourceStatus']
alasensitivelist

Unnamed: 0       id                                               name  \
0             0  2586820                  Tetratheca aphylla subsp. aphylla   
1             0  2589732                                 Tetratheca harperi   
2             0  2587391                                 Atriplex yeelirrie   
3             0  2589866             Tetratheca paynterae subsp. cremnobata   
4             0  2588285                                  Acacia shapelleae   
...         ...      ...                                                ...   
4199          0  2587070                                Leporillus conditor   
4200          0  2589588                     Phascogale tapoatafa wambenger   
4201          0  2587603  Bettongia lesueur subsp. (Barrow and Boodie Is...   
4202          0  2588671                                  Phascogale calura   
4203          0  2590930                      Cacatua pastinator pastinator   

                                 commonName  \
0                                       NaN   
1                        Jackson Tetratheca   
2                                       NaN   
3                                       NaN   
4                         Shapelle's Wattle   
...                                     ...   
4199                 Greater Stick-nest Rat   
4200  South-western Brush-tailed Phascogale   
4201     Boodie (barrow And Boodie Islands)   
4202                  Red-tailed Phascogale   
4203                         Muir's Corella   

                                       scientificName_x  \
0                     Tetratheca aphylla subsp. aphylla   
1                                    Tetratheca harperi   
2                                    Atriplex yeelirrie   
3                Tetratheca paynterae subsp. cremnobata   
4                                     Acacia shapelleae   
...                                                 ...   
4199                                Leporillus conditor   
4200                     Phascogale tapoatafa wambenger   
4201  Bettongia lesueur subsp. (Barrow and Boodie Is...   
4202                                  Phascogale calura   
4203           Cacatua (Licmetis) pastinator pastinator   

                                                   lsid dataResourceUid  \
0      https://id.biodiversity.org.au/node/apni/2902462         dr18406   
1      https://id.biodiversity.org.au/node/apni/2896021         dr18406   
2     https://id.biodiversity.org.au/taxon/apni/5127...         dr18406   
3      https://id.biodiversity.org.au/node/apni/2891052         dr18406   
4     https://id.biodiversity.org.au/taxon/apni/5128...         dr18406   
...                                                 ...             ...   
4199  https://biodiversity.org.au/afd/taxa/5ebc59f2-...         dr18406   
4200  https://biodiversity.org.au/afd/taxa/05cba699-...         dr18406   
4201                                    ALA_DR2201_4202         dr18406   
4202  https://biodiversity.org.au/afd/taxa/36b436b1-...         dr18406   
4203  https://biodiversity.org.au/afd/taxa/0e9219d1-...         dr18406   

                                              kvpValues W A Taxon Id  \
0     [{'key': 'W A Taxon Id', 'value': '29489'}, {'...        29489   
1     [{'key': 'W A Taxon Id', 'value': '4534'}, {'k...         4534   
2     [{'key': 'W A Taxon Id', 'value': '46173'}, {'...        46173   
3     [{'key': 'W A Taxon Id', 'value': '23987'}, {'...        23987   
4     [{'key': 'W A Taxon Id', 'value': '44470'}, {'...        44470   
...                                                 ...          ...   
4199  [{'key': 'status', 'value': 'Conservation Depe...          NaN   
4200  [{'key': 'status', 'value': 'Conservation Depe...          NaN   
4201  [{'key': 'status', 'value': 'Conservation Depe...          NaN   
4202  [{'key': 'status', 'value': 'Conservation Depe...          NaN   
4203  [{'key': 'status', 'value': 'Conservation Depe...          NaN   

                

In [106]:
statelist = pd.DataFrame(alasensitivelist[['waTaxonID','scientificName','status','wa_geoprivacy','lsid']])
statelist

waTaxonID                   scientificName                  status  \
0        29489       Tetratheca aphylla aphylla              Vulnerable   
1         4534               Tetratheca harperi              Vulnerable   
2        46173               Atriplex yeelirrie              Vulnerable   
3        23987  Tetratheca paynterae cremnobata              Vulnerable   
4        44470                Acacia shapelleae              Vulnerable   
...        ...                              ...                     ...   
4199       NaN              Leporillus conditor  Conservation Dependent   
4200       NaN   Phascogale tapoatafa wambenger  Conservation Dependent   
4201       NaN         Bettongia lesueur subsp.  Conservation Dependent   
4202       NaN                Phascogale calura  Conservation Dependent   
4203       NaN    Cacatua pastinator pastinator  Conservation Dependent   

     wa_geoprivacy                                               lsid  
0         obscured   https://id.biodiversity.org.au/node/apni/2902462  
1         obscured   https://id.biodiversity.org.au/node/apni/2896021  
2         obscured  https://id.biodiversity.org.au/taxon/apni/5127...  
3         obscured   https://id.biodiversity.org.au/node/apni/2891052  
4         obscured  https://id.biodiversity.org.au/taxon/apni/5128...  
...            ...                                                ...  
4199      obscured  https://biodiversity.org.au/afd/taxa/5ebc59f2-...  
4200      obscured  https://biodiversity.org.au/afd/taxa/05cba699-...  
4201      obscured                                    ALA_DR2201_4202  
4202      obscured  https://biodiversity.org.au/afd/taxa/36b436b1-...  
4203      obscured  https://biodiversity.org.au/afd/taxa/0e9219d1-...  

[4204 rows x 5 columns]

### 4. Equivalent IUCN statuses

In [107]:
iucn_statuses = {'Not Evaluated', 'Data Deficient', 'Least Concern', 'Near Threatened', 'Vulnerable', 'Endangered', 'Critically Endangered', 'Extinct in the Wild','Extinct'}
statelist.groupby(['status'])['status'].count()

status
Conservation Dependent                                                            7
Critically Endangered                                                           217
Endangered                                                                      198
Extinct                                                                          38
Other Specially Protected                                                         8
Priority                                                                        219
Priority 1 - Poorly known Species                                              1072
Priority 2 - Poorly known Species                                               835
Priority 3 - Poorly known species                                               981
Priority 4 - Rare , Near Threatened and other species in need of monitoring     366
Vulnerable                                                                      263
Name: status, dtype: int64

In [20]:
# map these statuses back to the original values
waStatusMappings = {
    'CD & MI': 'Special Conservation Interest',
    'MI & P1': 'P1 - Poorly known species',
    'MI & P3': 'P3 - Poorly known species',
    'Migratory & P4': 'P4 - Rare, Near Threatened and other species in need of monitoring',
    'P4 - Poorly known species': 'P4 - Rare, Near Threatened and other species in need of monitoring'
}
statelist['new_wa_status'] = statelist['status'].str.strip().map(waStatusMappings).fillna(statelist['status'])
#mergedstatuses['new_iucn_equivalent'] = mergedstatuses['status'].str.lower().str.strip().map(iucnStatusMappings).fillna('Vulnerable') # map to dictionary
statelist[['status','new_wa_status']].drop_duplicates()

status  \
0           P1 - Poorly known species   
1           P2 - Poorly known species   
2           P3 - Poorly known species   
8           P4 - Poorly known species   
16                   Threatened Flora   
131                  Presumed Extinct   
3859                       Vulnerable   
3864                        Migratory   
3866            Critically Endangered   
3886                       Endangered   
3899  Other specially protected fauna   
3943                          Extinct   
3945    Special Conservation Interest   
4216                          CD & MI   
4276                   Migratory & P4   
4330                          MI & P1   
4331                          MI & P3   

                                          new_wa_status  
0                             P1 - Poorly known species  
1                             P2 - Poorly known species  
2                             P3 - Poorly known species  
8     P4 - Rare, Near Threatened and other species i...  
16                                     Threatened Flora  
131                                    Presumed Extinct  
3859                                         Vulnerable  
3864                                          Migratory  
3866                              Critically Endangered  
3886                                         Endangered  
3899                    Other specially protected fauna  
3943                                            Extinct  
3945                      Special Conservation Interest  
4216                      Special Conservation Interest  
4276  P4 - Rare, Near Threatened and other species i...  
4330                          P1 - Poorly known species  
4331                          P3 - Poorly known species

In [8]:
iucnStatusMappings = {
    'critically endangered': 'Critically Endangered',
    'vulnerable':'Vulnerable',
    'not evaluated':'Not Evaluated',
    'data deficient':'Data Deficient',
    'least concern':'Least Concern',
    'special least concern':'Least Concern',
    'near threatened':'Near Threatened',
    'endangered':'Endangered',
    'extinct in the wild':'Extinct in the Wild',
    'extinct':'Extinct',
    'confidential':'Vulnerable'
}

### 5. Determine best place ID to use

In [9]:
inatstatuses.groupby(['place_id','place_name','place_display_name'])['place_id'].count()
# looks like 7308

place_id  place_name             place_display_name           
                                                                    5
144315    Brisbane City          Brisbane City                      2
153119    South East Queensland  South East Queensland, QL, AU      1
18870     Cairns - Pt B          Cairns - Pt B, QL, AU              1
19232     Yarrabah               Yarrabah, QL, AU                   1
6744      Australia              Australia                         12
7308      Queensland             Queensland, AU                   631
Name: place_id, dtype: int64

## Merge iNaturalist statuses with State lists on scientificName

1. Match - updates, even if the statuses are the same we'll update the links and values anyway
2. No match - statuses to be added (additions)
   1.1 No match and no taxnomy - search for synonyms
   1.2 No match
3. Merge the other direction to see if there are deletes?


In [21]:
# join to see which lists already have a status in inaturalist based on scientificName
mergedstatuses = statelist[['waTaxonID','scientificName','new_wa_status','wa_geoprivacy']].merge(inatstatuses[['status_id','scientificName','taxon_id','user_id','description','iucn','authority','status','geoprivacy','place_id','place_display_name']],how="left",left_on='scientificName',right_on='scientificName',suffixes=(None,'_inat')).sort_values(['scientificName'])
mergedstatuses


waTaxonID                                   scientificName  \
3860       NaN                          Abebaioscia troglodytes   
0      50593.0                          Abildgaardia pachyptera   
1      14112.0           Abutilon sp. Hamelin (A.M. Ashby 2196)   
2      14110.0      Abutilon sp. Onslow (F. Smith s.n. 10/9/61)   
3      43021.0  Abutilon sp. Pritzelianum (S. van Leeuwen 5095)   
...        ...                                              ...   
3857   29060.0                                  Zeuxine oblonga   
3856   29060.0                                  Zeuxine oblonga   
3858   36237.0                                  Zornia areolata   
3859   34478.0    Zornia sp. West Kimberley (C.A. Gardner 9942)   
4456       NaN                             Zyzomys pedunculatus   

                  new_wa_status wa_geoprivacy status_id taxon_id user_id  \
3860                 Vulnerable      obscured       NaN      NaN     NaN   
0     P1 - Poorly known species      obscured       NaN      NaN     NaN   
1     P2 - Poorly known species      obscured       NaN      NaN     NaN   
2     P3 - Poorly known species      obscured       NaN      NaN     NaN   
3     P3 - Poorly known species      obscured       NaN      NaN     NaN   
...                         ...           ...       ...      ...     ...   
3857  P2 - Poorly known species      obscured    153758   369267  708886   
3856  P2 - Poorly known species      obscured    169907   369267  702203   
3858  P1 - Poorly known species      obscured       NaN      NaN     NaN   
3859  P3 - Poorly known species      obscured       NaN      NaN     NaN   
4456      Critically Endangered      obscured    152903    45374  708886   

     description iucn                                       authority  \
3860         NaN  NaN                                             NaN   
0            NaN  NaN                                             NaN   
1            NaN  NaN                                             NaN   
2            NaN  NaN                                             NaN   
3            NaN  NaN                                             NaN   
...          ...  ...                                             ...   
3857               20  WA Department of Environment and Convservation   
3856               20                       Atlas of Living Australia   
3858         NaN  NaN                                             NaN   
3859         NaN  NaN                                             NaN   
4456               50  WA Department of Environment and Convservation   

                     status geoprivacy place_id     place_display_name  
3860                    NaN        NaN      NaN                    NaN  
0                       NaN        NaN      NaN                    NaN  
1                       NaN        NaN      NaN                    NaN  
2                       NaN        NaN      NaN                    NaN  
3                       NaN        NaN      NaN                    NaN  
...                     ...        ...      ...                    ...  
3857                     NT   obscured     6827  Western Australia, AU  
3856                     NT                6827  Western Australia, AU  
3858                    NaN        NaN      NaN                    NaN  
3859                    NaN        NaN      NaN                    NaN  
4456  critically endangered   obscured     6827  Western Australia, AU  

[4457 rows x 14 columns]

In [11]:
# prepare the export fields, common to New template and Update template
# new statuses
# Taxon_Name,Status,Authority,IUCN_equivalent,Description,iNaturalist_Place_ID,url,Taxon_Geoprivacy,Username,taxon_id
# updates
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username
mergedstatuses['new_authority'] = "WA Deparment of Biodiversity, Conservation and Attractions"
mergedstatuses['new_description'] = "Listed as Confidential - refer to https://www.data.qld.gov.au/dataset/queensland-confidential-species"
#mergedstatuses['new_url'] = "https://apps.des.qld.gov.au/species-search/details/?id=" + mergedstatuses['wildnetTaxonID'].astype(str)
florabaseurl = "https://florabase.dpaw.wa.gov.au/browse/profile/"
biesearchurl = "https://bie.ala.org.au/species/https://id.biodiversity.org.au/node/apni/2894366"
mergedstatuses['new_url'] = mergedstatuses.apply(lambda x: biesearchurl + x['lsid'] if pd.isna(x['waTaxonID']) else florabaseurl + x['waTaxonID'],axis=1)
floradescrurl = "Listed as Confidential - refer to https://www.dpaw.wa.gov.au/plants-and-animals/threatened-species-and-communities/threatened-plants"
faunadescrurl = "Listed as Confidential - refer to https://www.dpaw.wa.gov.au/plants-and-animals/threatened-species-and-communities/threatened-animals"
mergedstatuses['new_description'] = mergedstatuses.apply(lambda x: faunadescrurl if pd.isna(x['waTaxonID']) else floradescrurl,axis=1)
mergedstatuses.rename(columns={'wildnet_geoprivacy':'new_geoprivacy'},inplace=True)
mergedstatuses['new_place_id'] = '7308'  # Queensland, AU
mergedstatuses['new_username'] = 'peggydnew'
mergedstatuses['new_iucn_equivalent'] = mergedstatuses['status'].str.lower().str.strip().map(iucnStatusMappings).fillna('Vulnerable') # map to dictionary
mergedstatuses['new_status'] = mergedstatuses['status'].fillna('Confidential')
mergedstatuses

wildnetTaxonID        scientificName                 status  \
1676           33360   Abrodictyum brassii  Special least concern   
1677           33358  Abrodictyum caudatum  Special least concern   
1678           33359  Abrodictyum obscurum  Special least concern   
1769           11892    Acacia acrionastes        Near Threatened   
1770           14898      Acacia ammophila             Vulnerable   
...              ...                   ...                    ...   
2333           31680       Zieria scopulus  Critically Endangered   
2334           28217         Zieria vagans  Critically Endangered   
2335            3296      Zieria verrucosa             Vulnerable   
2336           41054    Zieria wilhelminae  Critically Endangered   
2525           14130    Zostera capricorni  Special least concern   

     new_geoprivacy status_id taxon_id user_id description iucn  \
1676           open       NaN      NaN     NaN         NaN  NaN   
1677           open       NaN      NaN     NaN         NaN  NaN   
1678           open       NaN      NaN     NaN         NaN  NaN   
1769           open       NaN      NaN     NaN         NaN  NaN   
1770           open       NaN      NaN     NaN         NaN  NaN   
...             ...       ...      ...     ...         ...  ...   
2333           open    168010  1097883   58320               50   
2334           open       NaN      NaN     NaN         NaN  NaN   
2335           open       NaN      NaN     NaN         NaN  NaN   
2336           open       NaN      NaN     NaN         NaN  NaN   
2525           open       NaN      NaN     NaN         NaN  NaN   

                                    authority  ... geoprivacy place_id  \
1676                                      NaN  ...        NaN      NaN   
1677                                      NaN  ...        NaN      NaN   
1678                                      NaN  ...        NaN      NaN   
1769                                      NaN  ...        NaN      NaN   
1770                                      NaN  ...        NaN      NaN   
...                                       ...  ...        ...      ...   
2333  Queensland Nature Conservation Act 1992  ...                7308   
2334                                      NaN  ...        NaN      NaN   
2335                                      NaN  ...        NaN      NaN   
2336                                      NaN  ...        NaN      NaN   
2525                                      NaN  ...        NaN      NaN   

     place_display_name                              new_authority  \
1676                NaN  Qld Department of Environment and Science   
1677                NaN  Qld Department of Environment and Science   
1678                NaN  Qld Department of Environment and Science   
1769                NaN  Qld Department of Environment and Science   
1770                NaN  Qld Department of Environment and Science   
...                 ...                                        ...   
2333     Queensland, AU  Qld Department of Environment and Science   
2334                NaN  Qld Department of Environment and Science   
2335                NaN  Qld Department of Environment and Science   
2336                NaN  Qld Department of Environment and Science   
2525                NaN  Qld Department of Environment and Science   

                                        new_description  \
1676  Listed as Confidential - refer to https://www....   
1677  Listed as Confidential - refer to https://www....   
1678  Listed as Confidential - refer to https://www....   
1769  Listed as Confidential - refer to https://www....   
1770  Listed as Confidential - refer to https://www....   
...                                                 ...   
2333  Listed as Confidential - refer to https://www....   
2334  Listed as Confidential - refer to https://www....   
2335  Listed as Confidential - refer to https://www....   
2336  Listed as Confidential - refer to https://www.

## Updates

In [12]:
# those that need to be updated - we found a status
mergedstatuses[mergedstatuses['status_id'].notnull()][['scientificName','status','status_inat','new_geoprivacy','geoprivacy','authority','user_id']]

scientificName                 status status_inat new_geoprivacy  \
1775      Acacia attenuata             Vulnerable          VU           open   
1776   Acacia barakulensis             Vulnerable          VU           open   
1777  Acacia baueri baueri             Vulnerable          VU           open   
1778       Acacia calantha        Near Threatened         NT            open   
1787     Acacia hockingsii             Vulnerable          VU           open   
...                    ...                    ...         ...            ...   
2318         Zieria bifida             Endangered          EN           open   
2321        Zieria collina             Vulnerable          VU           open   
2323          Zieria exsul  Critically Endangered          CR           open   
2326     Zieria gymnocarpa  Critically Endangered          CR           open   
2333       Zieria scopulus  Critically Endangered          CR           open   

     geoprivacy                                authority  user_id  
1775                                          Queensland   702203  
1776   obscured                                 QLD DEHP   702203  
1777                                          Queensland   702203  
1778                               Queensland Government   702203  
1787       open  Queensland Nature Conservation Act 1992  3669610  
...         ...                                      ...      ...  
2318             Queensland Nature Conservation Act 1992   702203  
2321       open             Nature Conservation Act 1992   702203  
2323                                          Queensland   702203  
2326                                            QLD DEHP   409010  
2333             Queensland Nature Conservation Act 1992    58320  

[570 rows x 7 columns]

In [13]:
# updates - create the data frame
# action,taxon_name,id,taxon_id,status,iucn_equivalent,authority,url,geoprivacy,place_id,username,description
updates = pd.DataFrame(mergedstatuses[mergedstatuses['status_id'].notnull()])
updates.sort_values('scientificName')
updates['action'] = 'UPDATE'
#updates.loc[:,'action'] = 'UPDATE'
updates = updates[['action','scientificName','status_id','taxon_id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
updates.columns = updates.columns.str.replace("new_", "", regex=True)
updates = updates.rename(columns={'scientificName':'taxon_name',
                                  'status_id':'id'})
updates

action            taxon_name      id taxon_id                 status  \
1775  UPDATE      Acacia attenuata  159918   898644             Vulnerable   
1776  UPDATE   Acacia barakulensis  158182  1038687             Vulnerable   
1777  UPDATE  Acacia baueri baueri  159921  1111700             Vulnerable   
1778  UPDATE       Acacia calantha  160798  1121183        Near Threatened   
1787  UPDATE     Acacia hockingsii  264938  1252519             Vulnerable   
...      ...                   ...     ...      ...                    ...   
2318  UPDATE         Zieria bifida  223195  1125311             Endangered   
2321  UPDATE        Zieria collina  262040   537414             Vulnerable   
2323  UPDATE          Zieria exsul  159923  1111558  Critically Endangered   
2326  UPDATE     Zieria gymnocarpa  169061  1244660  Critically Endangered   
2333  UPDATE       Zieria scopulus  168010  1097883  Critically Endangered   

            iucn_equivalent                                  authority  \
1775             Vulnerable  Qld Department of Environment and Science   
1776             Vulnerable  Qld Department of Environment and Science   
1777             Vulnerable  Qld Department of Environment and Science   
1778        Near Threatened  Qld Department of Environment and Science   
1787             Vulnerable  Qld Department of Environment and Science   
...                     ...                                        ...   
2318             Endangered  Qld Department of Environment and Science   
2321             Vulnerable  Qld Department of Environment and Science   
2323  Critically Endangered  Qld Department of Environment and Science   
2326  Critically Endangered  Qld Department of Environment and Science   
2333  Critically Endangered  Qld Department of Environment and Science   

                                                    url geoprivacy place_id  \
1775  https://apps.des.qld.gov.au/species-search/det...       open     7308   
1776  https://apps.des.qld.gov.au/species-search/det...       open     7308   
1777  https://apps.des.qld.gov.au/species-search/det...       open     7308   
1778  https://apps.des.qld.gov.au/species-search/det...       open     7308   
1787  https://apps.des.qld.gov.au/species-search/det...       open     7308   
...                                                 ...        ...      ...   
2318  https://apps.des.qld.gov.au/species-search/det...       open     7308   
2321  https://apps.des.qld.gov.au/species-search/det...       open     7308   
2323  https://apps.des.qld.gov.au/species-search/det...       open     7308   
2326  https://apps.des.qld.gov.au/species-search/det...       open     7308   
2333  https://apps.des.qld.gov.au/species-search/det...       open     7308   

       username                                        description  
1775  peggydnew  Listed as Confidential - refer to https://www....  
1776  peggydnew  Listed as Confidential - refer to https://www....  
1777  peggydnew  Listed as Confidential - refer to https://www....  
1778  peggydnew  Listed as Confidential - refer to https://www....  
1787  peggydnew  Listed as Confidential - refer to https://www....  
...         ...                                                ...  
2318  peggydnew  Listed as Confidential - refer to https://www....  
2321  peggydnew  Listed as Confidential - refer to https://www....  
2323  peggydnew  Listed as Confidential - refer to https://www....  
2326  peggydnew  Listed as Confidential - refer to https://www....  
2333  peggydnew  Listed as Confidential - refer to https://www....  

[570 rows x 12 columns]

In [14]:
# investigation - which updates are mine (346), which are those not from me (80 not from me)
#updates[updates['user_id']=='708886'][['scientificName','new_status','status_inat','authority','new_authority','description','new_description','geoprivacy','new_geoprivacy']]
#updates[updates['user_id']!='708886'][['user_id','scientificName','new_status','status_inat','authority','new_authority','description','new_description','geoprivacy','new_geoprivacy']]
# those with different statuses
#updates[updates['new_status'].str.lower().str.strip() != updates['status_inat'].str.lower().str.strip()][['scientificName','new_status','status_inat','authority','new_authority','description','new_description','geoprivacy','new_geoprivacy']]
# users who've updated qld statuses who aren't me
#'https://www.inaturalist.org/users/220795','Steven Kurniawidjaja','neontetraploid','US'
#'https://www.inaturalist.org/users/3669610','Craig Robbins','craig-r','AU'
#'https://www.inaturalist.org/users/527710','James Kameron Mitchell','jameskm','US'
#'https://www.inaturalist.org/users/58320','lwnrngr','lwnrngr','NZ'
#'https://www.inaturalist.org/users/702203','Kitty Maurey','kitty12','CA'
#'https://www.inaturalist.org/users/717122','Miguel de Salas','mftasp','TAS'


## No status in iNaturalist via straight scientificName match
The Qld records that didn't match up to a status in iNaturalist

In [15]:
# to add: those that have no inaturalist status - 532!!
noinatstatus = mergedstatuses[mergedstatuses['status_id'].isnull()]
# try to match the taxon name to something in inaturalist
noinatstatus = noinatstatus.merge(inattaxa, how="left", left_on="scientificName",right_on="scientificName")
noinatstatus

wildnetTaxonID        scientificName                 status  \
0              33360   Abrodictyum brassii  Special least concern   
1              33358  Abrodictyum caudatum  Special least concern   
2              33359  Abrodictyum obscurum  Special least concern   
3              11892    Acacia acrionastes        Near Threatened   
4              14898      Acacia ammophila             Vulnerable   
...              ...                   ...                    ...   
1956            9476       Zieria rimulosa             Vulnerable   
1957           28217         Zieria vagans  Critically Endangered   
1958            3296      Zieria verrucosa             Vulnerable   
1959           41054    Zieria wilhelminae  Critically Endangered   
1960           14130    Zostera capricorni  Special least concern   

     new_geoprivacy status_id taxon_id user_id description iucn authority  \
0              open       NaN      NaN     NaN         NaN  NaN       NaN   
1              open       NaN      NaN     NaN         NaN  NaN       NaN   
2              open       NaN      NaN     NaN         NaN  NaN       NaN   
3              open       NaN      NaN     NaN         NaN  NaN       NaN   
4              open       NaN      NaN     NaN         NaN  NaN       NaN   
...             ...       ...      ...     ...         ...  ...       ...   
1956           open       NaN      NaN     NaN         NaN  NaN       NaN   
1957           open       NaN      NaN     NaN         NaN  NaN       NaN   
1958           open       NaN      NaN     NaN         NaN  NaN       NaN   
1959           open       NaN      NaN     NaN         NaN  NaN       NaN   
1960           open       NaN      NaN     NaN         NaN  NaN       NaN   

      ...        phylum           class            order            family  \
0     ...           NaN             NaN              NaN               NaN   
1     ...  Tracheophyta  Polypodiopsida  Hymenophyllales  Hymenophyllaceae   
2     ...           NaN             NaN              NaN               NaN   
3     ...  Tracheophyta   Magnoliopsida          Fabales          Fabaceae   
4     ...  Tracheophyta   Magnoliopsida          Fabales          Fabaceae   
...   ...           ...             ...              ...               ...   
1956  ...  Tracheophyta   Magnoliopsida       Sapindales          Rutaceae   
1957  ...  Tracheophyta   Magnoliopsida       Sapindales          Rutaceae   
1958  ...  Tracheophyta   Magnoliopsida       Sapindales          Rutaceae   
1959  ...           NaN             NaN              NaN               NaN   
1960  ...  Tracheophyta      Liliopsida      Alismatales       Zosteraceae   

            genus specificEpithet infraspecificEpithet              modified  \
0             NaN             NaN                  NaN                   NaN   
1     Abrodictyum        caudatum                  NaN  2022-06-07T08:14:14Z   
2             NaN             NaN                  NaN                   NaN   
3          Acacia     acrionastes                  NaN  2022-04-06T22:12:50Z   
4          Acacia       ammophila                  NaN  2022-04-07T01:03:28Z   
...           ...             ...                  ...                   ...   
1956       Zieria        rimulosa                  NaN  2021-07-28T04:47:47Z   
1957       Zieria          vagans                  NaN  2021-07-28T04:47:46Z   
1958       Zieria       verrucosa                  NaN  2021-07-28T04:47:50Z   
1959          NaN             NaN                  NaN                   NaN   
1960      Zostera      capricorni                  NaN  2019-02-15T07:39:29Z   

     taxonRank                      references  
0          NaN                             NaN  
1      species                             NaN  
2          NaN                             NaN  
3      species    http://powo.science.kew.org/  
4      species    http://powo.science.kew.org/  
...        ...                             ...  
1956   spec

In [16]:
noinatstatus[noinatstatus['id'].notna()] # there's no status but there is a matching inat taxon (id is the taxon id)
# note: "Dendrobium" matches to both genus and section

wildnetTaxonID        scientificName                 status  \
1              33358  Abrodictyum caudatum  Special least concern   
3              11892    Acacia acrionastes        Near Threatened   
4              14898      Acacia ammophila             Vulnerable   
5               3304        Acacia arbiana        Near Threatened   
6              31097      Acacia argentina             Vulnerable   
...              ...                   ...                    ...   
1955           27791        Zieria obovata             Vulnerable   
1956            9476       Zieria rimulosa             Vulnerable   
1957           28217         Zieria vagans  Critically Endangered   
1958            3296      Zieria verrucosa             Vulnerable   
1960           14130    Zostera capricorni  Special least concern   

     new_geoprivacy status_id taxon_id user_id description iucn authority  \
1              open       NaN      NaN     NaN         NaN  NaN       NaN   
3              open       NaN      NaN     NaN         NaN  NaN       NaN   
4              open       NaN      NaN     NaN         NaN  NaN       NaN   
5              open       NaN      NaN     NaN         NaN  NaN       NaN   
6              open       NaN      NaN     NaN         NaN  NaN       NaN   
...             ...       ...      ...     ...         ...  ...       ...   
1955           open       NaN      NaN     NaN         NaN  NaN       NaN   
1956           open       NaN      NaN     NaN         NaN  NaN       NaN   
1957           open       NaN      NaN     NaN         NaN  NaN       NaN   
1958           open       NaN      NaN     NaN         NaN  NaN       NaN   
1960           open       NaN      NaN     NaN         NaN  NaN       NaN   

      ...        phylum           class            order            family  \
1     ...  Tracheophyta  Polypodiopsida  Hymenophyllales  Hymenophyllaceae   
3     ...  Tracheophyta   Magnoliopsida          Fabales          Fabaceae   
4     ...  Tracheophyta   Magnoliopsida          Fabales          Fabaceae   
5     ...  Tracheophyta   Magnoliopsida          Fabales          Fabaceae   
6     ...  Tracheophyta   Magnoliopsida          Fabales          Fabaceae   
...   ...           ...             ...              ...               ...   
1955  ...  Tracheophyta   Magnoliopsida       Sapindales          Rutaceae   
1956  ...  Tracheophyta   Magnoliopsida       Sapindales          Rutaceae   
1957  ...  Tracheophyta   Magnoliopsida       Sapindales          Rutaceae   
1958  ...  Tracheophyta   Magnoliopsida       Sapindales          Rutaceae   
1960  ...  Tracheophyta      Liliopsida      Alismatales       Zosteraceae   

            genus specificEpithet infraspecificEpithet              modified  \
1     Abrodictyum        caudatum                  NaN  2022-06-07T08:14:14Z   
3          Acacia     acrionastes                  NaN  2022-04-06T22:12:50Z   
4          Acacia       ammophila                  NaN  2022-04-07T01:03:28Z   
5          Acacia         arbiana                  NaN  2022-04-06T22:14:00Z   
6          Acacia       argentina                  NaN  2022-04-03T23:48:35Z   
...           ...             ...                  ...                   ...   
1955       Zieria         obovata                  NaN  2020-02-19T19:36:10Z   
1956       Zieria        rimulosa                  NaN  2021-07-28T04:47:47Z   
1957       Zieria          vagans                  NaN  2021-07-28T04:47:46Z   
1958       Zieria       verrucosa                  NaN  2021-07-28T04:47:50Z   
1960      Zostera      capricorni                  NaN  2019-02-15T07:39:29Z   

     taxonRank                      references  
1      species                             NaN  
3      species    http://powo.science.kew.org/  
4      species    http://powo.science.kew.org/  
5      species    http://powo.science.kew.org/  
6      species    http://powo.science.kew.org/  
...        ...                             ...  
1955   spec

In [17]:
# there's no status but there is a matching inat taxon (id is the taxon id)
additions = pd.DataFrame(noinatstatus[noinatstatus['id'].notna()])
additions.sort_values(['scientificName'])
additions['action'] = 'ADD'
additions = additions[['action','scientificName','status_id','id','new_status','new_iucn_equivalent','new_authority','new_url','new_geoprivacy','new_place_id','new_username','new_description']]
additions.columns = additions.columns.str.replace("new_", "", regex=True)
additions = additions.rename(columns={'scientificName':'taxon_name',
                                      'id':'taxon_id',
                                  'status_id':'id'})
additions

action            taxon_name   id taxon_id                 status  \
1       ADD  Abrodictyum caudatum  NaN   451374  Special least concern   
3       ADD    Acacia acrionastes  NaN   898576        Near Threatened   
4       ADD      Acacia ammophila  NaN   898600             Vulnerable   
5       ADD        Acacia arbiana  NaN   898626        Near Threatened   
6       ADD      Acacia argentina  NaN   898630             Vulnerable   
...     ...                   ...  ...      ...                    ...   
1955    ADD        Zieria obovata  NaN   956303             Vulnerable   
1956    ADD       Zieria rimulosa  NaN  1247621             Vulnerable   
1957    ADD         Zieria vagans  NaN  1247619  Critically Endangered   
1958    ADD      Zieria verrucosa  NaN  1247620             Vulnerable   
1960    ADD    Zostera capricorni  NaN   407068  Special least concern   

            iucn_equivalent                                  authority  \
1             Least Concern  Qld Department of Environment and Science   
3           Near Threatened  Qld Department of Environment and Science   
4                Vulnerable  Qld Department of Environment and Science   
5           Near Threatened  Qld Department of Environment and Science   
6                Vulnerable  Qld Department of Environment and Science   
...                     ...                                        ...   
1955             Vulnerable  Qld Department of Environment and Science   
1956             Vulnerable  Qld Department of Environment and Science   
1957  Critically Endangered  Qld Department of Environment and Science   
1958             Vulnerable  Qld Department of Environment and Science   
1960          Least Concern  Qld Department of Environment and Science   

                                                    url geoprivacy place_id  \
1     https://apps.des.qld.gov.au/species-search/det...       open     7308   
3     https://apps.des.qld.gov.au/species-search/det...       open     7308   
4     https://apps.des.qld.gov.au/species-search/det...       open     7308   
5     https://apps.des.qld.gov.au/species-search/det...       open     7308   
6     https://apps.des.qld.gov.au/species-search/det...       open     7308   
...                                                 ...        ...      ...   
1955  https://apps.des.qld.gov.au/species-search/det...       open     7308   
1956  https://apps.des.qld.gov.au/species-search/det...       open     7308   
1957  https://apps.des.qld.gov.au/species-search/det...       open     7308   
1958  https://apps.des.qld.gov.au/species-search/det...       open     7308   
1960  https://apps.des.qld.gov.au/species-search/det...       open     7308   

       username                                        description  
1     peggydnew  Listed as Confidential - refer to https://www....  
3     peggydnew  Listed as Confidential - refer to https://www....  
4     peggydnew  Listed as Confidential - refer to https://www....  
5     peggydnew  Listed as Confidential - refer to https://www....  
6     peggydnew  Listed as Confidential - refer to https://www....  
...         ...                                                ...  
1955  peggydnew  Listed as Confidential - refer to https://www....  
1956  peggydnew  Listed as Confidential - refer to https://www....  
1957  peggydnew  Listed as Confidential - refer to https://www....  
1958  peggydnew  Listed as Confidential - refer to https://www....  
1960  peggydnew  Listed as Confidential - refer to https://www....  

[1355 rows x 12 columns]

In [18]:
# what didnt match to a taxon?
noinatstatus[noinatstatus['id'].isna()]


wildnetTaxonID                                     scientificName  \
0              33360                                Abrodictyum brassii   
2              33359                               Abrodictyum obscurum   
9              41056                                    Acacia castorum   
14             40993                                    Acacia forsteri   
19             14862                                       Acacia lauta   
...              ...                                                ...   
1930            6296  Xanthorrhoea sp. (Cape Bedford M.Gandini AQ601...   
1937           41972                                    Xylosma craynii   
1940           41347                      Zealandia pustulata pustulata   
1942           40091                                  Zeuxine attenuata   
1959           41054                                 Zieria wilhelminae   

                     status new_geoprivacy status_id taxon_id user_id  \
0     Special least concern           open       NaN      NaN     NaN   
2     Special least concern           open       NaN      NaN     NaN   
9                Vulnerable           open       NaN      NaN     NaN   
14    Critically Endangered           open       NaN      NaN     NaN   
19               Vulnerable           open       NaN      NaN     NaN   
...                     ...            ...       ...      ...     ...   
1930  Special least concern           open       NaN      NaN     NaN   
1937             Vulnerable           open       NaN      NaN     NaN   
1940  Special least concern           open       NaN      NaN     NaN   
1942          Least concern       obscured       NaN      NaN     NaN   
1959  Critically Endangered           open       NaN      NaN     NaN   

     description iucn authority  ... phylum class order family genus  \
0            NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   
2            NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   
9            NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   
14           NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   
19           NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   
...          ...  ...       ...  ...    ...   ...   ...    ...   ...   
1930         NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   
1937         NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   
1940         NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   
1942         NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   
1959         NaN  NaN       NaN  ...    NaN   NaN   NaN    NaN   NaN   

     specificEpithet infraspecificEpithet modified taxonRank references  
0                NaN                  NaN      NaN       NaN        NaN  
2                NaN                  NaN      NaN       NaN        NaN  
9                NaN                  NaN      NaN       NaN        NaN  
14               NaN                  NaN      NaN       NaN        NaN  
19               NaN                  NaN      NaN       NaN        NaN  
...              ...                  ...      ...       ...        ...  
1930             NaN                  NaN      NaN       NaN        NaN  
1937             NaN                  NaN      NaN       NaN        NaN  
1940             NaN                  NaN      NaN       NaN        NaN  
1942             NaN                  NaN      NaN       NaN        NaN  
1959             NaN                  NaN      NaN       NaN        NaN  

[606 rows x 36 columns]

In [19]:
noinatstatus.groupby('status').size()

status
Critically Endangered    105
Endangered               223
Extinct                    5
Extinct in the wild       23
Least concern             92
Near Threatened          219
Special least concern    738
Vulnerable               465
dtype: int64

### are there any that need to be removed?
qld list count: 2517
qld inat statuses count: 653

updates to inat status: 570
additional inat status: 1355
qld statuses we can't find a taxon match for in iNaturalist: 606
total: 2531 (explainable via the various genus/section entries that we matched to in the taxonomy)

inat statuses left over: 653-570=83 that may need checking against the above

In [20]:
# inat statuses that aren't in added or updated
inatstatuses[~inatstatuses['taxon_id'].isin(updates['taxon_id'])]


status_id taxon_id  user_id place_id source_id  \
102     159922  1019990   702203     7308             
437     234929   104980  3669610     7308             
18      167722  1073668  3669610     7308             
308     264555  1074506  3669610     7308             
2614    161805  1091300   702203     7308             
...        ...      ...      ...      ...       ...   
897     154451   937266   708886     7308     16653   
837     154918   954413   708886     7308     16653   
2815    264331   961868  3669610     7308             
1053    155106   963748   708886     7308     16653   
23      167723   993605  3669610     7308             

                                             authority      status  \
102                                         Queensland          VU   
437            Queensland Nature Conservation Act 1992          VU   
18                               Queensland Government          LC   
308   Queensland Department of Environment and Science          LC   
2614                      Nature Conservation Act 1992          EN   
...                                                ...         ...   
897                                           QLD DEHP  endangered   
837                                           QLD DEHP  endangered   
2815           Queensland Nature Conservation Act 1992          LC   
1053                                          QLD DEHP  endangered   
23                                            QLD DEHP          NT   

                                                    url  \
102   https://www.legislation.qld.gov.au/view/html/i...   
437   https://apps.des.qld.gov.au/species-search/det...   
18    https://www.data.qld.gov.au/dataset/conservati...   
308   https://apps.des.qld.gov.au/species-search/det...   
2614  http://155.187.2.69/cgi-bin/sprat/public/publi...   
...                                                 ...   
897   https://data.qld.gov.au/dataset/conservation-s...   
837   https://data.qld.gov.au/dataset/conservation-s...   
2815  https://apps.des.qld.gov.au/species-search/det...   
1053  https://data.qld.gov.au/dataset/conservation-s...   
23    https://www.data.qld.gov.au/dataset/conservati...   

                                            description geoprivacy  ...  \
102                                                                 ...   
437                                                                 ...   
18    Least concern, locations still considered conf...             ...   
308                                                           open  ...   
2614                                                                ...   
...                                                 ...        ...  ...   
897                                                       obscured  ...   
837                                                       obscured  ...   
2815                                                          open  ...   
1053                                                      obscured  ...   
23                                                        obscured  ...   

               genus specificEpithet infraspecificEpithet  \
102           Acacia          baueri                        
437   Maccullochella          peelii                        
18           Liparis   coelogynoides                        
308       Eucalyptus     chloroclada                        
2614                                                        
...              ...             ...                  ...   
897                                                         
837                                                         
2815      Eucalyptus      thozetiana                        
1053        Nervilia         simplex                        
23         Acianthus   amplexicaulis                        

                  modified             scientificName taxonRank  \
102   2022-04-06T22:03:46Z              Acacia baueri   species   
437   2022-01-28T04:34:59Z